### DISCUSSION TITLE
Download only downloads files from first page.

### DISCUSSION THREAD

**Discussion by ILSua** [microsoft-search/pnp-modern-search#4046] [Q&A] [answered]
**Version used**
v4.13.1

**Describe the bug**
When downloading multiple files, across multiple pages, only the file(s) from the first page (or sometimes the page you are currently on) is downloaded.

**To Reproduce**
Using Details List layout.
'Allow items selection', Allow multiple selection', and 'Enable download' are all enabled/toggled on.
Show paging is enabled. Showing 10 items per page, and allow 5 pages in range.
Hide navigation buttons (prev page, next page) - toggled off
Hide first/last navigation buttons - toggled off
Hide navigation buttons (prev, next, first, last) if they are disabled - toggled on

**Expected behavior**
I would expect that if I select one file from page one of the results being displayed, and a file from page two of the results, then clicking download would download a zip with both files contained within.

**Desktop (please complete the following information):**
Using Chrome (Version 129.0.6668.90)

**Additional context**
You guys are legends. Appreciate this app and the versions that continue to improve!!

**Comment by patrikhellgren** [comment, marked as answer]
It will only download files that are selected on the current page. This is by design since the results layout only knows about the items on the current page and because of this the download button cannot know anything about the selected items on other pages.

**Comment by ILSua** [comment]
> It will only download files that are selected on the current page. This is by design since the results layout only knows about the items on the current page and because of this the download button cannot know anything about the selected items on other pages.

But if I select a file on page one, then navigate to page 2 and select a file, then navigate back to page 1, that file is still selected.
So it maintains my selection across multiple pages, but when clicking the download button, it will only download the file from the page that I am on when I click the download button.

**Comment by patrikhellgren** [comment]
The results webpart only knows about the items on the current page. When you go to another page the items are reset and it loads just the items to display the new current page. This is so that it shouldn’t keep to many items in memory at the same time and not load that much over the network in the same request.
All selected items (just the item index) are on the other hand kept for all pages.
This is just how it is designed and nothing to do about when using paging.
If you would like to download items from the whole result set in one go then you need to show the whole list on one page. You could enable sticky headers for the details list and set a fixed height for the layout instead of having multiple pages and then scroll through the list to select the ones to download.

**Reply by ILSua** [reply]
Thanks for the reply Patrik.
I understand that maintaining the information on selected file properties could slow loading due to memory used to retain those properties, but I feel this would be a somewhat common use case.
If we can retain the properties in memory for 20 files from one page, it seems logical we could retain the file properties of 10 files from 2 pages?

**Reply by patrikhellgren** [reply]
That simply is how paging is designed to work and is the standard behaviour in all REST APIs (SharePoint, Graph etc), i.e. just load one page over the network in each paged call. It is not just to save memory but also network traffic, DOM size etc. REST paging is there for a reason. If you are not aware of how the paged layouts work behind the scenes you could have a look at the network calls when you step through the pages and you will notice that the results returned for each page only contains the specific list items for the page you are currently at.

To load all (say 500 since that is the maximum and you must see beyond your example of 10x2) items over the network first and then display them in a paged details list without loading anything over the network later on would probably require a new details list layout which you are of course more than welcome to contribute to or you could develop your own layout extension that does what you want. Or this could be a feature request for someone else to pick up on.

As I said there is also already a workaround, use the sticky headers option to have all items on the same page but still keep the same format for the webpart.

**Reply by ILSua** [reply]
Also just to note, the single page type set up to enable downloading multiple files, it has a limit of 500 items per page.
I have occasional search results 800 items being retuned. So, this still has to be downloaded in a batches (of 2). Much better than batches of 10. But just noting that constraint.
